source/mainnet/tutorials/onReceivingCIS2/on-receivingCIS2.rst

```diff
@@ -0,0 +1,384 @@
+    }
+
+    impl From<CustomContractError> for Cis2Error<CustomContractError> {
+        fn from(c: CustomContractError) -> Self {
+            Cis2Error::Custom(c)
+        }
+    }
+
+    impl<T> From<CallContractError<T>> for CustomContractError {
+        fn from(_: CallContractError<T>) -> Self {
+            Self::InvokeContractError
+        }
+    }
+
+    type ContractError = Cis2Error<CustomContractError>;
+
+A simple message string will be stored in the state.
+
+3. Modify the initialization function:
+
+.. code-block:: rust
+
+    #[init(contract = "token_forwarder")]
+    fn init(_ctx: &InitContext, _state_builder: &mut StateBuilder) -> InitResult<State> {
+        Ok(State {
+            message: "Token Forwarder initialized!".to_string(),
+        })
+    }
+
+This method will initialize the contract and assign a value to the **message string**.
+
+4. Modify the token receiving function. This will become the main method of the contract, the ``onReceivingCIS2`` function that handles incoming tokens:
+
+.. code-block:: rust
+
+    #[receive(
+            contract = "token_forwarder",
+            name = "onReceivingCIS2",
+            parameter = "OnReceivingCis2Params<ContractTokenId, ContractTokenAmount>",
+            error = "ContractError"
+        )]
+        fn contract_receive_cis2(ctx: &ReceiveContext, host: &Host<State>) -> Result<(), ContractError> {
+            // Get information about received tokens
+            let params: OnReceivingCis2Params<ContractTokenId, ContractTokenAmount> = ctx.parameter_cursor().get()?;
+
+            // Get the token contract that sent the tokens
+            let token_contract = match ctx.sender() {
+                Address::Contract(contract) => contract,
+                _ => return Ok(()), // Non-contract senders are ignored
+            };
+
+            // Create transfer to forward tokens to contract owner
+            let transfer = Transfer {
+                token_id: params.token_id,
+                amount: params.amount,
+                from: Address::Contract(ctx.self_address()),
+                to: Receiver::from_account(ctx.owner()),
+                data: AdditionalData::empty(),
+            };
+
+            // Execute the transfer
+            host.invoke_contract_read_only(
+                &token_contract,
+                &TransferParams::from(vec![transfer]),
+                "transfer",
+                Amount::zero(),
+            )?;
+
+            Ok(())
+        }
+
+5. Add helper functions to manage the state:
+
+.. code-block:: rust
+
+    #[receive(
+        contract = "token_forwarder",
+        name = "set_message",
+        parameter = "String",
+        error = "ContractError",
+        mutable
+    )]
+    fn set_message(ctx: &ReceiveContext, host: &mut Host<State>) -> Result<(), ContractError> {
+        let new_message: String = ctx.parameter_cursor().get()?;
+        host.state_mut().message = new_message;
+        Ok(())
+    }
```

Comment:
To keep contract code simple, I feel we don't need this function around manipulating the message strings.